ath79: swap eth0/eth1 on some SoCs with builtin switch

981213 opened the pull request:
This makes gmac order the same as ar71xx so that we don't need a config migration script for this.
Doing so also helps merging incorrectly split mdio/gmac driver in the future.

Tested on:
TP-Link TL-WR841N v8
TP-Link TL-WR941N v4

981213 committed:
ath79: ag71xx: defer probe if of_phy_connect failed

gmac0 may need a phy on builtin switch, which can be unavailable
if gmac0 is probed before builtin switch.
Return -EPROBE_DEFER in this case so that gmac0 can be probed
later.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
ath79: dts: drop "simple-mfd" for gmacs in SoC dtsi

With a proper probe deferring for ag71xx we don't need to explicitly
probe mdio1 before gmac0.
Drop all "simple-mfd" in SoC dtsi so that gmac orders can be the same
as ar71xx.
This makes eth0/eth1 order the same as those in ar71xx, which means
we don't need a migration script for this anymore and we can merge
incorrectly split gmac/mdio driver back together.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
ath79: fix default config for devices with eth0/eth1 swapped

also fix the following problems in this commit:
glinet,gl-ar150: This router uses an uncommon order of setting up gmacs
                 in ar71xx. gmac0 is preferred to be wan port because of
                 the additional link status info available. So this
                 router will have eth0/eth1 swapped comparing to ar71xx.
tplink,tl-wr710n-v1: same as gl-ar150
embeddedwireless,dorin: eth0 is used as switch port, which was incorrect.
                        It's correct now, so keep this one untouched.
tplink,tl-wr842n-v1: we don't swap PHYs on ar7241 so the original port order
                     is incorrect.
reorder archer-a7-v5 entry.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

ynezz commented on target/linux/ath79/dts/ar7241.dtsi:
```diff
@@ -78,7 +78,7 @@
 };
 
 &eth1 {
-	compatible = "qca,ar7241-eth", "syscon", "simple-mfd";
```
@981213 shouldn't you remove this as well? https://github.com/openwrt/openwrt/blob/71ab2c9d179652d9b528817ec04d3f1dd3b4b1df/target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c#L1516-L1520

981213 replied:
This should be kept here because it's now the only way mdio driver get probed. It can be dropped after actually merging mdio and eth driver back together.

ynezz replied:
I mean, this in comment is no longer true or is it ? https://github.com/openwrt/openwrt/blob/71ab2c9d179652d9b528817ec04d3f1dd3b4b1df/target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c#L1512

981213 replied:
It's still true as I didn't drop the ability to use simple-mfd to probe mdio before eth.
As it'll be dropped later anyway I think it isn't necessary to use a commit only to drop/modify some comment lines.
Checking OF_POPULATED_BUS is still necessary because the driver doesn't unregister the mdio node after probe deferring. It should be kept just to prevent double registering of mdio device.

adrianschmutzler commented:
I do not understand what you are doing here in detail, but there are also "simple-mfd" entries in device DTS files, not only in the SoC DTSI files:

```
grep -rn "simple-mfd" target/linux/ath79/
src/openwrt/target/linux/ath79/files/drivers/net/ethernet/atheros/ag71xx/ag71xx_main.c:1512:     * by adding a "simple-mfd" compatible to gmac node. The
src/openwrt/target/linux/ath79/dts/qca9533_tplink_cpe210.dtsi:118:      compatible = "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar7241_ubnt_bullet-m.dts:12: compatible = "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar9330_pqi_air-pen.dts:139:  compatible = "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar9330.dtsi:195:     compatible = "qca,ar9330-eth", "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar9331_pisen_wmm003n.dts:102:        compatible = "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar7242.dtsi:48:      compatible = "qca,ar7242-eth", "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar7242.dtsi:75:      compatible = "qca,ar7242-eth", "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar9331_tplink_tl-mr3040-v2.dts:156:  compatible = "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar9331_tplink_tl-wr703n_tl-mr10u.dtsi:99:    compatible = "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar9331_tplink_tl-mr3020-v1.dts:171:  compatible = "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/qca953x.dtsi:277:    compatible = "qca,qca9530-eth", "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/qca956x.dtsi:260:    compatible = "qca,qca9560-eth", "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/qca956x.dtsi:303:    compatible = "qca,qca9560-eth", "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar7241_ubnt_unifi.dts:128:   compatible = "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ath79.dtsi:46:                       compatible = "qca,ath79-eth", "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ath79.dtsi:69:                       compatible = "qca,ath79-eth", "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar7241_ubnt_rocket-m.dts:12: compatible = "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar7241.dtsi:81:      compatible = "qca,ar7241-eth", "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar934x.dtsi:243:     compatible = "qca,ar9340-eth", "syscon", "simple-mfd";
src/openwrt/target/linux/ath79/dts/ar9331_etactica_eg200.dts:83:        compatible = "syscon", "simple-mfd";
```

Can those be left or do they have to be removed, too?

981213 commented:
@adrianschmutzler 

> Can those be left or do they have to be removed, too?

Those should be left untouched for now. They are there for using mdio bus without enabling corresponding eth block.

ynezz commented:
> I do not understand what you are doing here in detail

For the background leading to changes in this PR you can see comments in #1990. In short, ath79 has currently different probing order of mdio block then ar71xx on the devices which use internal switch, so this would need some migration script in order to make migration from ar71xx to ath79 seamless. In order to avoid the need for this script, this PR is trying to make probing in the same order as on ar71xx, so the migration script wouldn't be needed.

ynezz commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -209,37 +211,21 @@ ath79_setup_interfaces()
-		ucidef_add_switch "switch0" \
-			"0@eth1" "1:lan" "2:lan" "3:lan" "4:lan"
-		;;
-	tplink,tl-wr710n-v1)
```
from fcc716e21d7fa07c68bac8886cef122d35fbbc01

> Sysupgrade from `ar71xx` works without glitches. Network interfaces assigned for LAN and WAN ports are `eth1` and `eth0` respectively, what's consistent with `ar71xx` target.

adrianschmutzler replied:
This also looks like tl-wr710n-v1 has been removed but not added again.

ynezz commented:
@rogerpueyo can you please check this on your COMFAST CF-E120A v3 ?
@marcin1j  can you please check this on your TL-WR710N v1 and TL-WR842N/ND v1 ?

adrianschmutzler commented:
I might be able to test on GL AR-150 later this week.

rogerpueyo commented:
Sure, tomorrow! :+1:

marcin1j commented:
I'm going to test both of them this weekend.

My old OpenWRT development environment is gone and I have to prepare a new machine, hence the delay.

rogerpueyo commented:
Hi @981213 ,

It does not seem to work properly for the Comfast CF-120A-v3. Before:

```
root@OpenWrt:~# ifconfig -a
br-lan    Link encap:Ethernet  HWaddr 39:39:39:39:3B:40  
          inet addr:192.168.1.1  Bcast:192.168.1.255  Mask:255.255.255.0
          inet6 addr: fe80::42a5:efff:fe29:fb40/64 Scope:Link
          inet6 addr: fd90:f7ea:e47b::1/60 Scope:Global
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1481 errors:0 dropped:0 overruns:0 frame:0
          TX packets:132 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:273702 (267.2 KiB)  TX bytes:21818 (21.3 KiB)

eth0      Link encap:Ethernet  HWaddr 39:39:39:39:3B:40  
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1482 errors:0 dropped:0 overruns:0 frame:0
          TX packets:137 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:294599 (287.6 KiB)  TX bytes:22244 (21.7 KiB)
          Interrupt:4 

eth1      Link encap:Ethernet  HWaddr 39:39:39:39:3B:42  
          inet addr:10.1.94.76  Bcast:10.1.94.95  Mask:255.255.255.224
          inet6 addr: fd00:1714:1714:2032:42a5:efff:fe29:fb42/64 Scope:Global
          inet6 addr: fd00:1714:1714:2032::676/128 Scope:Global
          inet6 addr: fe80::42a5:efff:fe29:fb42/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:6779 errors:0 dropped:0 overruns:0 frame:0
          TX packets:700 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:1393211 (1.3 MiB)  TX bytes:50884 (49.6 KiB)
          Interrupt:5 
[...]
```
With your branch:
```
root@OpenWrt:~# ifconfig -a
br-lan    Link encap:Ethernet  HWaddr 39:39:39:39:3B:42  
          inet addr:192.168.1.1  Bcast:192.168.1.255  Mask:255.255.255.0
          inet6 addr: fe80::42a5:efff:fe29:fb42/64 Scope:Link
          inet6 addr: fdc8:d15b:c527::1/60 Scope:Global
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:15 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:1970 (1.9 KiB)

eth0      Link encap:Ethernet  HWaddr 39:39:39:39:3B:42  
          inet6 addr: fe80::42a5:efff:fe29:fb42/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:1541 errors:0 dropped:0 overruns:0 frame:0
          TX packets:303 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:188454 (184.0 KiB)  TX bytes:96098 (93.8 KiB)
          Interrupt:5 

eth0.1    Link encap:Ethernet  HWaddr 39:39:39:39:3B:42  
          inet6 addr: fe80::42a5:efff:fe29:fb42/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:192 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:60960 (59.5 KiB)

eth0.2    Link encap:Ethernet  HWaddr 39:39:39:39:3B:42  
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:15 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:1970 (1.9 KiB)

eth1      Link encap:Ethernet  HWaddr 39:39:39:39:3B:40  
          BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:4
[...]
```
/etc/config/network:
```
config interface 'lan'
	option type 'bridge'
	option proto 'static'
	option ipaddr '192.168.1.1'
	option netmask '255.255.255.0'
	option ip6assign '60'
	option ifname 'eth0.2'

config interface 'wan'
	option proto 'dhcp'
	option ifname 'eth0.1'

config interface 'wan6'
	option proto 'dhcpv6'
	option ifname 'eth0.1'
```
I don't know why it uses eth0.1 and eth0.2, if they're not specified anywhere. Is it maybe not going through the default case in 02_network? If I edit the config/network file and set eth0 and eth1 for lan and wan, respectively, it works well.

981213 commented:
I just noticed that I completely forgot to swap stuff in 01_leds :(
@rogerpueyo I guess cf-e120a-v3 got a broken wan. If led configuration for cf-e120a-v3 is correct, the wan port is on phy1 instead of phy0.

I'm fixing these problems.

981213 commented:
rebased and fixed the mentioned problem.

rogerpueyo commented:
@981213 
I ended up having a weird 02_network file that was not properly rebased, I guess, on my compilation directory. It was my fault; I tested later, compiling from scratch on a different directory and the network was working fine. :+1: 
I'll test the LEDs tomorrow, I don't have physical access to the device now.

981213 commented:
@rogerpueyo 

> I tested later, compiling from scratch on a different directory and the network was working fine. 

What? It should have a broken WAN port (or leds defined in 01_leds is incorrect.)

> I'll test the LEDs tomorrow, I don't have physical access to the device now.

Could you test if LED for WAN port is working correct? If not, could you plug a cable into wan port and paste the output of "swconfig dev switch0 show" command?

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -62,27 +62,27 @@ engenius,ecb1750)
 	;;
 glinet,gl-ar150)
-	ucidef_set_led_netdev "wan" "WAN" "$boardname:green:wan" "eth0"
+	ucidef_set_led_netdev "wan" "WAN" "$boardname:green:wan" "eth1"
```
AR150: WAN is linked to eth1 here, but in 02-network it is eth0?

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -71,15 +84,11 @@ ath79_setup_interfaces()
-	yuncore,a770)
+	ubnt,nanostation-m|\
+	ubnt,routerstation)
 		ucidef_set_interfaces_lan_wan "eth1" "eth0"
```
WAN is eth0 here, but eth1 for LEDs. During a quick test on the AR150, WAN-Port was working and LAN-Port was "strange".

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -71,15 +84,11 @@ ath79_setup_interfaces()
 		ucidef_add_switch "switch0" \
 			"0@eth0" "1:lan:1" "3:lan:4" "4:lan:3" "5:lan:2" "2:wan"
 		;;
-	comfast,cf-e5|\
```
cf-e5 is deleted, but not recreated

981213 replied:
I meant to do this. It defaults to the last case in this script.

adrianschmutzler replied:
Ah, I see. I always thought of the default case as an emergency fallback, but why not use it ...

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -71,15 +84,11 @@ ath79_setup_interfaces()
+	comfast,cf-e120a-v3|\
 	glinet,gl-ar150|\
-	glinet,gl-ar300m-nand|\
-	glinet,gl-ar300m-nor|\
```
gl-ar300m-nand and glinet,gl-ar300m-nor are only deleted, but not recreated.

981213 replied:
Both of them also uses the the last default case.

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -49,9 +49,9 @@ ath79_setup_interfaces()
 	tplink,archer-c58-v1|\
 	tplink,archer-c59-v1)
-		ucidef_set_interface_wan "eth0"
+		ucidef_set_interface_wan "eth1"
```
WAN does not match LEDS for C58/59.

981213 replied:
fixed

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -203,6 +195,17 @@ ath79_setup_interfaces()
+	tplink,archer-d50-v1)
+		ucidef_add_switch "switch0" \
+			"0@eth0" "2:lan:3" "3:lan:2" "4:lan:1" "1:wan"
+		;;
```
WAN LEDs need update from eth1.2 to eth0.2.

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -62,7 +62,20 @@ ath79_setup_interfaces()
+	tplink,tl-wr841-v9|\
+	tplink,tl-wr841-v11|\
+	tplink,tl-wr842n-v1|\
+	tplink,tl-wr842n-v3|\
```
Semi-related question: 841-v7 and 842n-v1 do not have a LED-setup at all (unrelated to this patch).
Is the desired?

981213 replied:
AR7241 has working hardware switch LEDs so it doesn't need a software one.

adrianschmutzler replied:
Okay :-)

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -71,15 +84,11 @@ ath79_setup_interfaces()
-	tplink,tl-wr810n-v1|\
-	tplink,tl-wr810n-v2|\
-	ubnt,routerstation|\
-	yuncore,a770)
```
This whole removal block is not recreated, except ubnt,routerstation below. (rebase issue?)

981213 replied:
All of these uses the last default case.

adrianschmutzler commented:
Hi, just went through the whole 01_leds and 02_network files and compared how ethX are used there. Unfortunately, it looks like there has been a rebase issue or something similar, since some devices got lost. (see inline comments above)

981213 commented:
@adrianschmutzler 
What's that "strange" behavior on GL-AR150? I assume it got wan/lan swapped and some weird LED activity but I can't think of anything else.

I just noticed that in ar71xx, GL-AR150 uses a uncommon way of registering both gmacs, and it has the same eth order as current ath79 one (without this PR.)
I think I may have to break sysupgrade compatibility from ar71xx to ath79 for this router because:
With this PR, gmac1 becomes eth0 and gmac0 becomes eth1.
Using gmac1 as WAN port isn't a good idea as we lost the ability to detect link change on WAN port and behave accordingly.

adrianschmutzler commented:
@981213
I only had time for a quick test on AR150 and did not get an IP address on client port. However, from a quick look into config everything looked fine. Since this has been done quickly between two other things, though, I will have to look at it more thoroughly before I can give proper feedback.

981213 committed:
ath79: rework LED configurations for tplink,archer-d50-v1

The original one has the following problem:
1. Port mask of lan led includes wan port.
2. By using netdev trigger with vlan port, the link led
   is always on.

This commits fixes the above problems by correcting port
mask for lan led and use swconfig trigger for wan leds.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 commented:
@adrianschmutzler 
On GL-AR150, wan is eth1 now. Could you test with my latest commits next time?

adrianschmutzler commented:
@981213 
I plan on testing AR150 today or tomorrow evening

marcin1j commented:
I've tested this PR on TL-WR710N and TL-WR842ND. It swaps LAN and WAN network interfaces as intended.

On TL-WR842ND everything (LAN and WAN interfaces, default config) works fine.

However on TL-WR710N WAN interface is unusable. It won't detect network link (WAN is `eth1`):
```
[    0.744964] ag71xx 19000000.eth: Could not connect to PHY device. Deferring probe.
[    1.479726] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    1.488269] eth0: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
[    1.855553] ag71xx 19000000.eth: connected to PHY at mdio-bus.0:1f:04 [uid=004dd041, driver=Generic PHY]
[    1.864992] eth1: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
[    7.026657] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[    8.094995] eth0: link up (1000Mbps/Full duplex)
[    8.098197] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   11.178627] eth0: link down
[   29.121494] IPv6: ADDRCONF(NETDEV_UP): eth1: link is not ready
[   29.169769] br-iot: port 1(eth1.7) entered blocking state
[   29.173723] br-iot: port 1(eth1.7) entered disabled state
[   29.179866] device eth1.7 entered promiscuous mode
[   29.183942] device eth1 entered promiscuous mode
[   29.760578] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[   29.766268] eth0: link up (1000Mbps/Full duplex)
[   29.770109] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   29.854451] br-lan: port 1(eth0.1) entered blocking state
[   29.858406] br-lan: port 1(eth0.1) entered disabled state
[   29.864545] device eth0.1 entered promiscuous mode
[   29.868557] device eth0 entered promiscuous mode
[   29.954005] br-lan: port 1(eth0.1) entered blocking state
[   29.957966] br-lan: port 1(eth0.1) entered forwarding state
[   30.461407] br-lan: port 2(eth1.1) entered blocking state
[   30.465447] br-lan: port 2(eth1.1) entered disabled state
[   30.494015] device eth1.1 entered promiscuous mode
```

Building firmware from master branch boot log looks like (WAN is `eth0`):
```
[    1.425565] ag71xx 19000000.eth: connected to PHY at mdio-bus.0:1f:04 [uid=004dd041, driver=Generic PHY]
[    1.434606] eth0: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
[    1.775901] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    1.784412] eth1: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
[    7.689869] IPv6: ADDRCONF(NETDEV_UP): eth1: link is not ready
[    7.694470] IPv6: ADDRCONF(NETDEV_UP): eth1.1: link is not ready
[    8.735065] eth1: link up (1000Mbps/Full duplex)
[    8.738273] IPv6: ADDRCONF(NETDEV_CHANGE): eth1: link becomes ready
[    8.745177] IPv6: ADDRCONF(NETDEV_CHANGE): eth1.1: link becomes ready
[   19.016597] eth1: link down
[   38.430815] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[   38.472543] br-iot: port 1(eth0.7) entered blocking state
[   38.476588] br-iot: port 1(eth0.7) entered disabled state
[   38.482587] device eth0.7 entered promiscuous mode
[   38.486722] device eth0 entered promiscuous mode
[   39.453290] IPv6: ADDRCONF(NETDEV_UP): eth1: link is not ready
[   39.460041] eth1: link up (1000Mbps/Full duplex)
[   39.464399] IPv6: ADDRCONF(NETDEV_CHANGE): eth1: link becomes ready
[   39.528070] br-lan: port 1(eth1.1) entered blocking state
[   39.532072] br-lan: port 1(eth1.1) entered disabled state
[   39.537955] device eth1.1 entered promiscuous mode
[   39.542177] device eth1 entered promiscuous mode
[   39.600688] br-lan: port 1(eth1.1) entered blocking state
[   39.604719] br-lan: port 1(eth1.1) entered forwarding state
[   40.941132] br-lan: port 2(eth0.1) entered blocking state
[   40.945175] br-lan: port 2(eth0.1) entered disabled state
[   40.985881] device eth0.1 entered promiscuous mode
[  148.735490] eth0: link up (100Mbps/Full duplex)
[  148.738611] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[  148.750001] br-lan: port 2(eth0.1) entered blocking state
[  148.754042] br-lan: port 2(eth0.1) entered forwarding state
[  148.760153] br-iot: port 1(eth0.7) entered blocking state
[  148.764940] br-iot: port 1(eth0.7) entered forwarding state
```

981213 commented:
@marcin1j 
> It won't detect network link (WAN is eth1)
Could you plug in a network cable for both ports and paste the output of "swconfig dev switch0 show" command?

rogerpueyo commented:
@981213 COMFAST CF-E120A v3 working fine, LEDs included. :+1:

marcin1j commented:
> Could you plug in a network cable for both ports and paste the output of "swconfig dev switch0 show" command?

I'll check it tomorrow (don't have physical access to the device right now).

adrianschmutzler commented:
Tested successfully on GL-AR150. WAN is eth1, LAN is eth0 (swapped compared to ar71xx).
I also tested this with my PRs for Archer C25 und C60:
WAN is eth1, Switch is eth0, also swapped compared to ar71xx (as for the C58/C59 already existing in ath79).

marcin1j commented:
@981213 
Surprisingly, WAN interface now seems connected to internal switch port 1. LAN interface is connected to port 3 and CPU to port 0 as before.
If you take that into account and adjust config accordingly, WAN interface works as `eth0` VLAN.

Of course this is not compatible with `ar71xx` target. There's also a question of mixing tagged and untagged traffic on single ethernet port, which definitely was possible on separate WAN separate interface and AFAIK is impossible with `AR724X/AR933X` switch driver.

Just for the record, here's output of `swconfig dev switch0 show`:
```
Global attributes:
        enable_vlan: 1
        enable_mirror_rx: 0
        enable_mirror_tx: 0
        mirror_monitor_port: 0
        mirror_source_port: 0
        arl_table: address resolution table
Port 0: MAC XX:XX:XX:XX:XX:d5
Port 0: MAC XX:XX:XX:XX:XX:cb

Port 0:
        mib: MIB counters
RxBroad     : 110
RxPause     : 0
RxMulti     : 24
RxFcsErr    : 0
RxAlignErr  : 0
RxRunt      : 0
RxFragment  : 0
Rx64Byte    : 110
Rx128Byte   : 15
Rx256Byte   : 9
Rx512Byte   : 0
Rx1024Byte  : 0
Rx1518Byte  : 0
RxMaxByte   : 0
RxTooLong   : 0
RxGoodByte  : 9716 (9.4 KiB)
RxBadByte   : 0
RxOverFlow  : 0
Filtered    : 119
TxBroad     : 16
TxPause     : 0
TxMulti     : 10
TxUnderRun  : 0
Tx64Byte    : 15
Tx128Byte   : 16
Tx256Byte   : 3
Tx512Byte   : 8
Tx1024Byte  : 0
Tx1518Byte  : 0
TxMaxByte   : 0
TxOverSize  : 0
TxByte      : 5316 (5.1 KiB)
TxCollision : 0
TxAbortCol  : 0
TxMultiCol  : 0
TxSingleCol : 0
TxExcDefer  : 0
TxDefer     : 0
TxLateCol   : 0

        pvid: 0
        link: port:0 link:up speed:1000baseT full-duplex txflow rxflow 
Port 1:
        mib: MIB counters
RxBroad     : 94
RxPause     : 0
RxMulti     : 11
RxFcsErr    : 0
RxAlignErr  : 0
RxRunt      : 0
RxFragment  : 0
Rx64Byte    : 55
Rx128Byte   : 54
Rx256Byte   : 4
Rx512Byte   : 8
Rx1024Byte  : 0
Rx1518Byte  : 0
RxMaxByte   : 0
RxTooLong   : 0
RxGoodByte  : 10642 (10.3 KiB)
RxBadByte   : 0
RxOverFlow  : 0
Filtered    : 79
TxBroad     : 0
TxPause     : 0
TxMulti     : 7
TxUnderRun  : 0
Tx64Byte    : 0
Tx128Byte   : 7
Tx256Byte   : 0
Tx512Byte   : 0
Tx1024Byte  : 0
Tx1518Byte  : 0
TxMaxByte   : 0
TxOverSize  : 0
TxByte      : 614
TxCollision : 0
TxAbortCol  : 0
TxMultiCol  : 0
TxSingleCol : 0
TxExcDefer  : 0
TxDefer     : 0
TxLateCol   : 0

        pvid: 0
        link: port:1 link:up speed:100baseT full-duplex auto
Port 2:
        mib: No MIB data
        pvid: 0
        link: port:2 link:down
Port 3:
        mib: MIB counters
RxBroad     : 0
RxPause     : 0
RxMulti     : 0
RxFcsErr    : 0
RxAlignErr  : 0
RxRunt      : 0
RxFragment  : 0
Rx64Byte    : 0
Rx128Byte   : 0
Rx256Byte   : 0
Rx512Byte   : 0
Rx1024Byte  : 0
Rx1518Byte  : 0
RxMaxByte   : 0
RxTooLong   : 0
RxGoodByte  : 0
RxBadByte   : 0
RxOverFlow  : 0
Filtered    : 0
TxBroad     : 8
TxPause     : 0
TxMulti     : 0
TxUnderRun  : 0
Tx64Byte    : 8
Tx128Byte   : 0
Tx256Byte   : 0
Tx512Byte   : 0
Tx1024Byte  : 0
Tx1518Byte  : 0
TxMaxByte   : 0
TxOverSize  : 0
TxByte      : 512
TxCollision : 0
TxAbortCol  : 0
TxMultiCol  : 0
TxSingleCol : 0
TxExcDefer  : 0
TxDefer     : 0
TxLateCol   : 0

        pvid: 1
        link: port:3 link:up speed:100baseT full-duplex auto
Port 4:
        mib: No MIB data
        pvid: 0
        link: port:4 link:down
VLAN 1:
        vid: 1
        ports: 0t 3 
```

981213 commented:
@marcin1j 
Oops, that's my fault. I've fixed it now.
Due to the same reason as gl-ar150, wr710n-v1 will have eth0/eth1 swapped comparing to ar71xx.

marcin1j commented:
> @marcin1j
> Due to the same reason as gl-ar150, wr710n-v1 will have eth0/eth1 swapped comparing to ar71xx.

So there's no way to fix this? How did you make WAN interface appear connected to port 1?

kofec commented:
There is a way to fix it. Just profiles for  gl-ar150, wr710n-v1 should not be removed and old rules should be applied

adrianschmutzler commented:
The question is whether migration should really be the no. 1 priority for the single time where you might be doing it.
IMHO it would be desirable to set the "new" ath79 target up with the cleanest solution, unless introducing special cases just because of migration.

ynezz commented:
> The question is whether migration should really be the no. 1 priority for the single time where you might be doing it.

It was agreed upon, that its not just a priority, but basically a must have, it's de facto a rule https://openwrt.org/unsupported/ath79_wip and http://lists.infradead.org/pipermail/openwrt-devel/2018-December/014888.html

981213 commented:
@kofec 
> There is a way to fix it. Just profiles for gl-ar150, wr710n-v1 should not be removed and old rules should be applied

The main purpose of this PR is to help merging mdio and gmac driver back together. Compatibility to ar71xx is a 'side-effect' of it. The already-upstreamed ag71xx driver has made mdio and gmac driver together and we'll need ethernet ports ordered as gmac1=eth0 gmac0=eth1 in order to use the upstream driver anyway.

The problem of those two routers is caused by unstandard initialization sequence made by the author submitting ar71xx support. I don't think we should carry a mistake from ar71xx into ath79 just to get a sysupgrade compatibility.

kofec commented:
I'm not an expert at all. I just answer @marcin1j question - that it is possible :-). I didn't know that the changes are related to mdio and gmac drivers

ynezz commented:
@981213 I've now following on 8devices Carambola

ar71xx:

```
libphy: Fixed MDIO Bus: probed
libphy: ag71xx_mdio: probed
ag71xx-mdio.1: Found an AR7240/AR9330 built-in switch
eth0: Atheros AG71xx at 0xba000000, irq 5, mode:GMII
ag71xx ag71xx.0: connected to PHY at ag71xx-mdio.1:04 [uid=004dd041, driver=Generic PHY]
eth1: Atheros AG71xx at 0xb9000000, irq 4, mode:MII
```

ath79
```
libphy: Fixed MDIO Bus: probed
libphy: ag71xx_mdio: probed
ag71xx 19000000.eth: Could not connect to PHY device. Deferring probe.
libphy: ag71xx_mdio: probed
libphy: ar8xxx-mdio: probed
switch0: Atheros AR724X/AR933X built-in rev. 2 switch registered on mdio-bus.1
ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
eth0: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
ag71xx 19000000.eth: connected to PHY at mdio-bus.1:1f:04 [uid=004dd041, driver=Generic PHY]
eth1: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
```

>  I don't think we should carry a mistake from ar71xx into ath79 just to get a sysupgrade compatibility.

I agree. This two devices could be probably handled by the migration script, right?

ynezz commented:
Thanks, going to push it to master soon.

CodeFetch commented:
> The question is whether migration should really be the no. 1 priority for the single time where you might be doing it.
> IMHO it would be desirable to set the "new" ath79 target up with the cleanest solution, unless introducing special cases just because of migration.

@ynezz @981213 

I don't really understand what you are doing here. I would be very grateful if you could explain it in a simple way. At the moment e.g. on QCA9531 eth0 is the NIC with the switch attached while eth1 is the one without the switch. From my understanding it would make sense the other way around, because devices which only have a single port likely use the NIC without switch and this should be eth0. Furthermore most devices (if not all) use the single-port NIC for WAN and it is much more intuitive for the user if eth0 is WAN and eth1 is LAN.

981213 commented:
@CodeFetch 
There are two ethernet blocks on ath79 SoCs with builtin switch. gmac0 is the one with a PHY attached and gmac1 is the one connected to builtin switch. However this PHY used by gmac0 belongs to builtin switch and can only be accessed through mdio bus on builtin switch which is controlled with mdio bus on gmac1. We have to initialize gmac1 before gmac0 so that we could have access to builtin switch and that PHY used by gmac0.

In ar71xx everything are mixed up. Two ethernet blocks are reset by machine code, not the driver, and ath79-mdio driver is controlling switch mdio through gmac1 mdio, and that's why the order in machine code doesn't matter: You can use mdio before initializing corresponding ethernet block and you can control switch mdio without loading switch driver. In order to let drivers doing only what it should do, this specific initialization order is required.

981213 commented:
@CodeFetch 

>  If the upstream ag71xx driver forces this initialization order, it would make sense to fix it, doesn't it?

Yes.

> If you don't want to change anything here, can you please explain how to disable the NIC which has the switch attached for devices which only have got one port and use the NIC without switch only like KuWfi CPE2G-9531?

Change the compatible string of eth1 to
```
compatible = "syscon", "simple-mfd";
```
I incorrectly split eth and mdio driver apart before, and this change will load only mdio driver for eth1.

Implementing a correct solution for this is a different topic...

CodeFetch commented:
@981213 Hm... I don't really understand this. Maybe you can help...

E.g. QCA9531 has two GMACs, one internal switch and five connections for fast ethernet PHYs. The switch has 6 MACs. MAC0 is always connected to GMAC1 using a (G)MII interface. MAC2-4 are connected to the PHY lines 1-3 (and I also think MAC1 is connected to PHY4 sometimes). 

GMAC0 has a MII interface and is connected to MAC5 of the switch. It can either be configured to directly connect to PHY0 or PHY4 while the remaining PHYs are connected to the switch or it can be configured to only be connected to MAC5 while all PHYs are connected to the switch. 
Is my description correct?
So why does the switch need to be initialized at all when GMAC0 is configured to be directly connected to PHY0 or PHY4 which is often the case for single-port routers?

> Change the compatible string of eth1 to
> 
> ```
> compatible = "syscon", "simple-mfd";
> ```

The problem with this device (KuWfi CPE2g-9531) is that eth0 must be disabled and not eth1, because the PHY which is connected on the board is GMAC0. BTW shouldn't also `status = "disabled"` be set here?

I really think GMAC0 should be eth0 and GMAC1 eth1 to simplify this case...

CodeFetch commented:
> I incorrectly split eth and mdio driver apart before, and this change will load only mdio driver for eth1.

Wait... When I set this this for eth1 this means that eth0 won't be initialized? And what's the resulting name of the single remaining device in this case? Does this mean I will have only one ethernet device which is called "eth1"? If yes, I didn't know this is possible and makes me think again that GMAC0 should be eth0 and GMAC1 eth1...

981213 commented:
@CodeFetch 

> Is my description correct?

This is correct, but...

> So why does the switch need to be initialized at all when GMAC0 is configured to be directly connected to PHY0 or PHY4 which is often the case for single-port routers?

mii interface is just data link. PHY control interface, which is mdio, is connected to builtin switch, not gmac0.

981213 commented:
> Wait... When I set this this for eth1 this means that eth0 won't be initialized?

Sorry for the confusion. Here I refer to gmac1 because it is named 'eth1' in dts. You could find an example in dts for PISEN WMB001N

CodeFetch commented:
@981213 Thank you for your explanations. I better understand how GMAC0 is connected to the PHYs and why initialization of GMAC1 is needed. Still I don't really understand the DTS configuration. Some hopefully simple questions: Is GMAC0 always eth0 and GMAC1 eth1 in DTS? Why are on a running system on some routers GMAC0 eth1 and GMAC1 eth0 and on others with the same SoC they aren't? Is there a default for switch-phy-swap defined or better why does it need to be set to 1 on some routers and 0 on others and no default defined/what is the default behavior?

jow- commented:
> Furthermore most devices (if not all) use the single-port NIC for WAN and it is much more intuitive for the user if eth0 is WAN and eth1 is LAN.

I'm used to the other way around, eth0 (first interface) is LAN, eth1 is WAN. This was the situation on the few ar71xx devices I have used and it also mirrors the behavior of x86 as well. The first found nic there is LAN, the second - if it exists - will be designated WAN.

CodeFetch commented:
@jow- 

My reasons for thinking eth0 is more intuitive for WAN are:

- The WAN port is left from the LAN ports on every router I've seen so far. One could say it's a cultural prejudice that numeric order has to be from left to right, but LAN ports are also labeled from left to right on most devices.
- Most "client" devices only have a single ethernet port which is some kind of WAN and thus always eth0.
- Most routers have a single WAN port, but many LAN ports. When there isn't an internal switch, the device has several LAN NICs and one WAN NIC. It's very strange when eth0-eth7 are LAN NICs and the last one/eth8 is WAN.

jow- commented:
> The WAN port is left from the LAN ports on every router I've seen so far. One could say it's a cultural prejudice that numeric order has to be from left to right, but LAN ports are also labeled from left to right on most devices.

Just from looking at two devices I happen to have lying on my desk here:
- Xiaomi Mi WiFi R3G, the order is LAN1, LAN2, WAN
- Cisco EA3500, the order is LAN1, LAN2, LAN3, LAN4, WAN

> Most "client" devices only have a single ethernet port which is some kind of WAN and thus always eth0.

The way I see it is that most client devices have a sole ethernet port which receives the *internal* (LAN) configuration. Of course you could plug your laptops ethernet directly into the WAN modem, but that is not the common case I suppose.

> Most routers have a single WAN port, but many LAN ports. When there isn't an internal switch, the device has several LAN NICs and one WAN NIC. It's very strange when eth0-eth7 are LAN NICs and the last one/eth8 is WAN. (and it's even more strange when one adds another NIC and e.g. eth9 is a LAN NIC again)

I think such hardware configurations are usually seen on x86 boxes where it is quite simple for users to renumber/repurpose their eth netdevs. On DSA based devices, the netdevs will have names like "lan1", "lan2" or "wan", so the ordering there is not important.

Your suggestion of always denoting eth0 to be WAN also complicates common failsafe considerations. One cannot rely on the assumption that the first nic is a LAN one but different code paths have to be taken depending on the amount of interfaces present and one cannot rely on the *last* interface to be a LAN one either because an USB ethernet device might be plugged or something like ipheth might be loaded...

CodeFetch commented:
@jow- I see... What aside from WAN/LAN-intuitivity is still very confusing and was the point I actually joined this conversation is that DTS eth0 often differs from run-time eth0. This is documented nowhere and the knowledge one can gather even from reading all related commit messages is also very limited. I don't grasp why GMAC0 (eth0 in DTS) can't always be eth0 in a run-time environment.

adrianschmutzler commented:
Well, can't we just rename eth0 and eth1 in dts to something more distinctive from eth0/eth1 on running devices? This would only be cosmetic, but it would be the simplest sed job and it would at least ease the confusion a little.

Neustradamus commented:
Merged commit is:
- https://github.com/openwrt/openwrt/commit/c3a85189181827c8d5c2ab736428be30e4c13128
